### ISSUE TITLE
[dataio] [shovelio] merge functionality of I3File and I3FrameSequence (Trac #1431)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1423]
Both I3File and I3FrameSequence provide a C++ and Python interface for accessing frames in files. They should be merged at some point.

Migrated from https://code.icecube.wisc.edu/ticket/1431
```json
{
    "status": "closed", 
    "changetime": "2019-02-13T14:14:44", 
    "description": "Both I3File and I3FrameSequence provide a C++ and Python interface for accessing frames in files. They should be merged at some point.", 
    "reporter": "hdembinski", 
    "cc": "david.schultz, nega, cweaver, olivas", 
    "resolution": "fixed", 
    "_ts": "1550067284370534", 
    "component": "combo core", 
    "summary": "[dataio] [shovelio] merge functionality of I3File and I3FrameSequence", 
    "priority": "normal", 
    "keywords": "", 
    "time": "2015-11-11T20:00:22", 
    "milestone": "", 
    "owner": "david.schultz", 
    "type": "task"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20151111T20:04:14: david.schultz commented:

> Don't forget source:IceTray/projects/dataio/trunk/python/fileadaptor.py
> 
> It's yet another wrapper around I3File.

**Comment by icecube-trac** [comment]
Trac update at 20151111T20:16:27: 
* hdembinski commented:

> A comparison of I3File (I3File.h), I3SequentialFile (pybindings/I3File.cxx) aka I3File in Python, and I3FrameSequence (shovelio)
> 
> Common
> - reading compressed and uncompressed files
> - random access (frame index is generated behind the scenes)
> 
> Differences
> - I3File (I3File.h)
>   - single file
>   - single threaded
>   - specific keys may be skipped
> 
> - I3SequentialFile aka I3File in Python (pybindings/I3File.cxx)
>   - single file
>   - single threaded
>   - convenient interface for sequential access (more, pop_frame and friends)
>   - support for writing frames
> 
> - I3FrameSequence (shovelio)
>   - multiple files are represented as one continuous stream of frames
>   - multi threaded: frame index is generated in a separate thread
>   - specific keys may be skipped
>   - reading from pipe/socket (limited random access emulated through caching)


* hdembinski changed _comment2 from:

> A comparison of I3File (I3File.h), I3SequentialFile (pybindings/I3File.cxx) aka I3File in Python, and I3FrameSequence (shovelio)
> 
> Common
> - reading compressed and uncompressed files
> - random access (frame index is generated behind the scenes)
> - specific keys may be skipped
> 
> Differences
> - I3File (I3File.h)
>   - single file
>   - single threaded
> 
> - I3SequentialFile aka I3File in Python (pybindings/I3File.cxx)
>   - single file
>   - single threaded
>   - convenient interface for sequential access (more, pop_frame and friends)
>   - support for writing frames
> 
> - I3FrameSequence (shovelio)
>   - multiple files are represented as one continuous stream of frames
>   - multi threaded: frame index is generated in a separate thread
>   - reading from pipe/socket (limited random access emulated through caching)

to:

> 1447273959857033


* hdembinski changed _comment1 from:

> A comparison of I3File and I3FrameSequence
> 
> Common
> - reading compressed and uncompressed files
> - random access (frame index is generated behind the scenes)
> - specific keys may be skipped
> 
> Differences
> - I3File from dataio
>   - sequential access
>   - single file
>   - single threaded
> 
> - I3FrameSequence from shovelio
>   - random access
>   - multiple files are represented as one continuous stream of frames
>   - multi threaded: frame index is generated in a separate thread
>   - reading from pipe/socket (limited random access emulated through caching)

to:

> 1447273846591111


* hdembinski changed _comment0 from:

> A comparison of I3File and I3FrameSequence
> 
> Common
> - reading compressed and uncompressed files
> - random access (frame index is generated behind the scenes)
> - specific keys may be skipped
> 
> Differences
> - I3File from dataio
>   - sequential access
>   - single file
>   - single threaded
>   - support for writing frames
> 
> - I3FrameSequence from shovelio
>   - random access
>   - multiple files are represented as one continuous stream of frames
>   - multi threaded: frame index is generated in a separate thread
>   - reading from pipe/socket (limited random access emulated through caching)

to:

> 1447273072599853

**Comment by icecube-trac** [comment]
Trac update at 20160330T14:56:38: 
* olivas changed status from "new" to "accepted"
* olivas changed owner from "" to "olivas"

**Comment by icecube-trac** [comment]
Trac update at 20170220T18:27:20: 
* olivas changed status from "accepted" to "closed"
* olivas changed resolution from "" to "wontfix"
* olivas commented:

> I think this is a great suggestion, but not likely to happen.  I'm tempted to set the milestone to Long-Term Future, but that's really just a nice way of saying 'wontfix.'  So I'm just going to close this.

**Comment by icecube-trac** [comment]
Trac update at 20170516T21:54:42: 
* david.schultz commented:

> Reopen for #1913.
> 
> At this time, I'd be happy with just eliminating `dataio.fileadaptor` and moving shovelio out of steamshovel so it can be used as a replacement.


* david.schultz changed status from "closed" to "reopened"
* david.schultz changed resolution from "wontfix" to ""

**Comment by icecube-trac** [comment]
Trac update at 20170516T21:55:39: david.schultz changed milestone from "May 2016" to "July 2017"

**Comment by icecube-trac** [comment]
Trac update at 20170525T15:42:53: david.schultz commented:

> Writing some of my observations:
> 
> * icetray/open
>   * The actual raw files, used by I3Reader/I3Writer and all of the below.
> * I3File
>   * Only used in dataio-shovel.
> * I3SequentialFile
>   * The python binding for `dataio.I3File`, so used occasionally.
>   * Could technically replace as long as the python API stays the same.
> * fileadaptor.py
>   * Only used in dataio-pyshovel.
> * I3FrameSequence
>   * Only used in steamshovel.

**Comment by icecube-trac** [comment]
Trac update at 20170525T16:45:06: 
* david.schultz changed owner from "olivas" to "david.schultz"
* david.schultz commented:

> My current plan is to remove I3File, I3SequentialFile, and fileadaptor.py.  This will be replaced by the contents of shovelio, which can provide a new base I3File with the same interface as I3SequetialFile (for the python API), as well as I3FrameSequence which the shovels should use.


* david.schultz changed _comment0 from "My current plan is to remove I3File, I3SequentialFile, and fileadaptor.py.  This will be replaced by the contents of shovelio, which can provide a new base I3File with the same interface, as well as I3FrameSequence which the shovels should use." to "1495731047124758"
* david.schultz changed status from "reopened" to "accepted"

**Comment by icecube-trac** [comment]
Trac update at 20170629T21:17:24: 
* david.schultz changed resolution from "" to "fixed"
* david.schultz commented:

> In changeset 2927:
> 
> ```text
> #!CommitTicketReference repository="IceTray" revision="2927"
> merge c++ and pybinding I3File. eliminate python fileadaptor. merge shovelio I3FrameSequence. integrate I3FrameSeqeuence into dataio-pyshovel, allowing shoveling over multiple files. closes #1423. closes #1913
> ```


* david.schultz changed status from "accepted" to "closed"

**Comment by icecube-trac** [comment]
Trac update at 20190111T21:55:24: nwhitehorn commented:

> In changeset 169219:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="169219"
> merge c++ and pybinding I3File. eliminate python fileadaptor. merge shovelio I3FrameSequence. integrate I3FrameSeqeuence into dataio-pyshovel, allowing shoveling over multiple files. closes #1423. closes #1913
> ```

**Comment by icecube-trac** [comment]
Trac update at 20190213T14:14:44: 
* olivas commented:

> Milestone July 2017 deleted


* olivas changed milestone from "July 2017" to ""
